nil/core-concepts/shards-parallel-execution.mdx

```diff
@@ -29,6 +35,92 @@ As the number of execution shards grows, this process may be changed so that sha
+
+1. Contract 1 calls a function inside Contract 2, which creates a new outgoing message
+2. Shard 2 must copy this message and place it inside a block
+3. When Shard 2 starts processing messages from the block, Contract 2 executes the function that Contract 1 called
+
+For an in-depth explanation of how to perform and manage async calls, [**click here**](../getting-started/essentials/handling-async-execution).
+
+:::info
+
+All shards 'listen' to the latest blocks created by other shards. When creating a block, a shard must copy messages from other shards that have this shard as its destination. These messages do not go through the mempool and, therefore, cannot be lost.
+
+:::
+
+:::info[Callback pattern]
+
+There is also a variation of the above flow in which the function inside Contract 2 makes an async call back to Contract 1.
+
+In this case, the flow would work as follows:
+
+1. Contract 1 calls a function inside Contract 2, creating a new outgoing message copied by Shard 2
+2. Shard 2 processes the message, and Contract 2 calls the requested function
+3. As part of executing the function, Contract 2 makes an async call back to Contract 1
+4. Shard 1 picks up the new message from Shard 2 and executes it
+5. Contract 1 performs whatever operation was requested by Contract 2
+
+As shown above, the async call inside Contract 2 effectively becomes a callback function for Contract 1. Note that the use of this pattern is only possible if Contract 1 or Contract 2 share the same owner (or their owners cooperate with each other).
+
+:::
+
+### Internal vs. external messages 
+
+In =nil;, messages are divided into internal and external.
+
+Internal messages cover all types of async communication between smart contracts on the network regardless of the shard where they are deployed. When Contract 1 (deployed on Shard 1) calls a function in Contract 2 (deployed on Shard 2), Shard 2 spawns an internal message. Note that sync calls between contracts do not produce any messages.
+
+External messages can originate only in one case: when any entity that exists outside of the cluster (e.g., a dApp) sends a message to any smart contract on the network.
+
+:::info[Payment for messages]
+
+For internal messages, the message itself is the 'payer' entity. An internal message spends its `value` to cover processing costs. Note that the message only pays for the initial call to a 'top-level' smart contract. If the message produces a chain of sync calls, the message will only pay for the first call in the chain.
+
+For external messages, the receiving contract is the 'payer' entity. To process external messages, the receiving contract must implement a special function that uses arbitrary logic to verify these messages. If this function is absent, the contract will only accept internal messages. A more detailed explanation of how a contract may receive external messages is [**given here**](../getting-started/essentials/receiving-ext-messages).
```

Comment:
`A more detailed explanation is given in the documentation on [receiving external messages](...).`